KEP 2535: Ensure Secret Pulled Images (#49886)
kubernetes/website: content/en/docs/concepts/containers/images.md

This can be used to preload certain images for speed or as an alternative to authenticating to a
private registry.

All pods will have read access to any pre-pulled images, unless the 
[KubeletEnsureSecretPulledImages](#kubeletensuresecretpulledimages) feature gate
is enabled.

> #### Pre-pulled image authentication {#prepulledimageauthentication}

Reviewer: Actually, I was just reading the docs page again, I think you had the location correct, though I would move it a bit further under "Specifying imagePullSecrets on a Pod"

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

If the `KubeletEnsureSecretPulledImages` 

> [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)

Reviewer: nit: I'm not sure if it's common to link to the feature gate docs in this context, I think we could probably drop the link here

is enabled, Kubernetes will require valid authentication credentials for every pod

> image, even if that image is already present on the node. This is controlled by the

Reviewer: That's not very accurate, the feature is much more complicated than that, we need to be more specific here.

`ImagePullcredentialsVerificationPolicy` option in the
[Kubelet config](/docs/reference/config-api/kubelet-config.v1beta1#ImagePullCredentialsVerificationPolicy).

This configuration controls when image pull credentials must be verified if the
image is already present on the node. The `NeverVerify` setting mimics the behavior

> of having this feature gate disabled, while the `NeverVerifyPreloadedImages` setting

Reviewer: The enumaration of possible values should be a list to make it a bit easier to consume.

There are 4 different forms of the policy, I see 2 being documented here.

is the default behavior when the feature gate is enabled. The default behavior allows
images already present in the node, but not those previously pulled by the Kubelet, to
be pulled without verification. All other images, including those pulled by other pods,
will have their image pull credentials verified against the registry. Image pulls

> which reuse the same credentials (in the case of a restarted container, for example)

Reviewer: We'll want to document the definition of `same credentials` .

#### Ensure Image Pull Credential Verification {#ensureimagepullcredentialverification}

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

If the `KubeletEnsureSecretPulledImages` feature gate is enabled, Kubernetes
will validate image credentials for every image that requires credentials 
to be pulled, even if that image is already present on the node. This validation
ensures that images in a pod request which have not been successfully pulled with the
provided credentials must re-pull the images from the registry.
Additionally, image pulls that re-use the same credentials which previously resulted 
in a successful image pull will not need to re-pull from the registry and are instead 
validated locally without accessing the registry (provided the image is available locally).
This is controlled by the`imagePullCredentialsVerificationPolicy` field in the
[Kubelet configuration](/docs/reference/config-api/kubelet-config.v1beta1#ImagePullCredentialsVerificationPolicy).

This configuration controls when image pull credentials must be verified if the
image is already present on the node:

>  * `NeverVerify`: Mimics the behavior of having this feature gate disabled. If the image
>  is present locally, image pull credentials are not verified.

Reviewer 2: We use sembr (https://sembr.org) style of line breaks.
Also, we recommend indenting all lines in a list item. I.e.

```suggestion
 * `NeverVerify`: Mimics the behavior of having this feature gate disabled.
   If the image is present locally, image pull credentials are not verified.
```

 * `NeverVerifyPreloadedImages`: Images pulled outside the kubelet are not verified,
 but all other images will have their credentials verified. This is the default behavior.
 * `NeverVerifyAllowListedImages`: Images pulled outside the kubelet and mentioned within the 
 `preloadedImagesVerificationAllowlist` specified in the kubelet config are not verified.
 * `AlwaysVerify`: All images will have their credentials verified before they can
 be used.

This verification applies to [pre-pulled images](#pre-pulled-images), images pulled
using node-wide secrets, and images pulled using pod-level secrets.

{{< note >}}

> In the case of credential rotation, the credentials used to previously pull the image

Reviewer 2: ```suggestion
In the case of credential rotation, the credentials previously used to pull the image
```

> will continue to verify without accessing the registry. New or rotated credentials

Reviewer: ```suggestion
will continue to verify without the need to access the registry. New or rotated credentials
```

`imagePullSecrets` must all be in the same namespace as the Pod. The referenced
Secrets must be of type `kubernetes.io/dockercfg` or `kubernetes.io/dockerconfigjson`.

#### Ensure Image Pull Credential Verification {#ensureimagepullcredentialverification}

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

If the `KubeletEnsureSecretPulledImages` feature gate
is enabled, Kubernetes will require valid authentication credentials for every pod
image, even if that image is already present on the node. This is controlled by the
`ImagePullcredentialsVerificationPolicy` option in the

> [Kubelet config](/docs/reference/config-api/kubelet-config.v1beta1#ImagePullCredentialsVerificationPolicy).

Reviewer: ```suggestion
[Kubelet configuration](/docs/reference/config-api/kubelet-config.v1beta1#ImagePullCredentialsVerificationPolicy).
```

This configuration controls when image pull credentials must be verified if the
image is already present on the node:

>  * `NeverVerify`: Mimics the behavior of having this feature gate disabled. If the image

Reviewer: Let's stick to the language from the KEP and use "credential verification" instead of authentication.

I think we can just make the description of the policies from https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md#proposal a bit more human-readable and we'd be fine.

 be used.

With this feature enabled, image pulls which reuse the same credentials and image digest
(in the case of a restarted container, for example), will not need to reauthenticate to
the registry if that image is present either through caching or as a pre-pulled image.
The kubelet verifies this by comparing the secret reference and a hash of the credentials used
against the digest of the iamge being requested. If the reference, hash, and digest match a
previously successful image pull, and a copy of the image is available locally either
through pre-pulling or caching, then no call will be made to the registry and the local
image will be used. Otherwise, the kubelet will authenticate the image pull against the
registry before allowing the pod to run the image, even if the image is already present.
This verification applies to [pre-pulled images](#pre-pulled-images), images pulled using
node-wide secrets, and images pulled using pod-level secrets.

{{< note >}}
Image pull credentials which expire will successfully verify if the same credential and image
digest are used. If expiring pull credentials are a concern, use an `ImagePullPolicy` of

> `Always` to always force kubelet to authenticate against the registry.

Reviewer: I think it's good to mention that we won't force credential verification if the credentials inside of a secret rotate.

I think rather than mentioning "expiring" credentials, I would just mention that the feature accounts for credential rotation, in which case reverifying the credentials with the registry may not be required.

> {{< /note >}}

Reviewer: If we manage to polish the initial description so that it clearly states when an image pull is required to verify credentials and when it isn't, we may not need this description.

A thing to note: image digest don't really matter much for the feature. Let's maybe use "image name from pod spec" and "container runtime image reference" when necessary?

 * `NeverVerifyAllowListedImages`: Images pulled outside the kubelet and mentioned within the 
 `preloadedImagesVerificationAllowlist` specified in the kubelet config are not verified.
 * `AlwaysVerify`: All images will have their credentials verified before they can
 be used.

With this feature enabled, image pulls which reuse the same credentials and image
(in the case of a restarted container, for example), will not need to re-pull from
the registry if that image is present either through caching or as a pre-pulled image.
The kubelet verifies this by comparing the secret reference and a hash of the credentials used
against the CRI image reference being requested. If the reference, hash, and image match a
previously successful image pull, and a copy of the image is available locally either
through pre-pulling or caching, then no call will be made to the registry and the local
image will be used. Otherwise, the kubelet will use the provided credentials to re-pull
the image from the registry before allowing the pod to run the image, even if the image is
already present. This verification applies to [pre-pulled images](#pre-pulled-images),

> images pulled using node-wide secrets, and images pulled using pod-level secrets.

Reviewer: I think this is a good description now, though I'd suggest dropping it as the intro to the section summarizes the higher level of the feature appropriately enough.
